### PR TITLE
All Category videos component layout

### PR DESCRIPTION
**Pull request by eldiegod**
implements: #1362

Basic UI Layout for All Category videos component.

Details on the search bar to be discussed with designers as it possibly will be removed (it is removed)

Filters and popover to come in a different PR

### CODE REVIEW

File: src/components/CategoryVideos/CategoryVideos.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+  const handleFilterClick = () => {
+    setiIsFiltersOpen((value) => !value)
+  }
+  return (
+    <Container>
+      <ControlsContainer>
+        <Text variant={mdMatch ? 'h4' : 'h5'}>All videos (441)</Text>
+        {mdMatch && <Select helperText={null} placeholder="Language" value={null} items={SORT_OPTIONS} />}
```
Comment (drillprop) [inline src/components/CategoryVideos/CategoryVideos.tsx]:
```suggestion
        {mdMatch && <Select size="small" helperText={null} placeholder="Language" value={null} items={SORT_OPTIONS} />}
```

File: src/components/CategoryVideos/CategoryVideos.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+          ) : (
+            <Button variant="secondary" icon={<SvgActionFilters />} onClick={handleFilterClick}>
+              Filters
+            </Button>
+          )}
+        </FiltersSearchContainer>
+        {mdMatch && (
+          <SortContainer>
+            <Text variant="body2">Sort by</Text>
+            <Select helperText={null} value={sortVideosBy} items={SORT_OPTIONS} onChange={handleSorting} />
```
Comment (drillprop) [inline src/components/CategoryVideos/CategoryVideos.tsx]: It's just looks really weird right now: 
image


```suggestion
            <Select size="small" helperText={null} value={sortVideosBy} items={SORT_OPTIONS} onChange={handleSorting} />
```

File: src/components/CategoryVideos/CategoryVideos.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    (searchQuery: string) => {
+      setSearchQuery(searchQuery)
+      searchVideo({
+        variables: {
+          text: searchQuery,
+          whereVideo: {
+            isPublic_eq: true,
+            mediaAvailability_eq: AssetAvailability.Accepted,
+            thumbnailPhotoAvailability_eq: AssetAvailability.Accepted,
+            // channelId_eq: '0',
```
Comment (drillprop) [inline src/components/CategoryVideos/CategoryVideos.tsx]:
```suggestion
```

File: src/components/CategoryVideos/CategoryVideos.styles.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import styled from '@emotion/styled'
+
+import { colors, media, sizes } from '@/shared/theme'
+
+export const Container = styled.div`
+  margin: ${sizes(16)} 0 0 0;
```
Comment (drillprop) [inline src/components/CategoryVideos/CategoryVideos.styles.tsx]:
```suggestion
  margin-top: ${sizes(16)};
```

File: src/components/CategoryVideos/CategoryVideos.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        <Text variant={mdMatch ? 'h4' : 'h5'}>All videos (441)</Text>
+        {mdMatch && <Select helperText={null} placeholder="Language" value={null} items={SORT_OPTIONS} />}
+        <FiltersSearchContainer>
+          <Search
+            searchInputRef={searchInputRef}
+            isSearchInputOpen={isSearchInputOpen}
+            setIsSearchingInputOpen={setIsSearchingInputOpen}
+            setIsSearching={setIsSearching}
+            search={search}
+          />
```
Comment (drillprop) [inline src/components/CategoryVideos/CategoryVideos.tsx]: Search input should not expand to the full width.
Design:
image

File: src/components/CategoryVideos/CategoryVideos.tsx (hunk 1 above)
Comment (drillprop) [inline src/components/CategoryVideos/CategoryVideos.tsx]: Also why do we have here `Language` as a placeholder and `SORT_OPTIONS` as items?

File: src/hooks/useToggle.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+export const useToggle = (initialState = false) => {
+  // Initialize the state
+  const [state, setState] = useState(initialState)
+
+  // Define and memorize toggler function in case we pass down the comopnent,
+  // This function change the boolean value to it's opposite value
+  const toggle = useCallback(() => setState((state) => !state), [])
+
+  return [state, toggle] as const
```
Comment (drillprop) [inline src/hooks/useToggle.ts]:
```suggestion
  return [state, toggle]
```
I don't think we need this

File: src/hooks/useToggle.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { useCallback, useState } from 'react'
+
+export const useToggle = (initialState = false) => {
+  // Initialize the state
```
Comment (drillprop) [inline src/hooks/useToggle.ts]:
```suggestion
```
I think we all know what it does 😅

File: src/hooks/useToggle.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  const [state, setState] = useState(initialState)
+
+  // Define and memorize toggler function in case we pass down the comopnent,
+  // This function change the boolean value to it's opposite value
```
Comment (drillprop) [inline src/hooks/useToggle.ts]:
```suggestion
```
same

File: src/shared/components/Search/Search.tsx [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        setIsSearchingInputOpen(false)
+        searchInputRef.current?.blur()
+        setSearchQuery('')
+      }
+    },
+    [search, searchInputRef, searchQuery, setIsSearching, setIsSearchingInputOpen, setSearchQuery]
+  )
+
+  return (
+    <SearchContainer className={className}>
```
Comment (drillprop) [inline src/shared/components/Search/Search.tsx]: I said this before in other PR and I still maintain this. We should just use `form` here, like this:
```tsx
    <SearchContainer
      as="form"
      className={className}
      onSubmit={(e) => {
        e.preventDefault()
        // all on submit logic like search(searchQuery)
      }}
    >
```
Hitting `enter` will just work out of the box(but we could leave listening on `escape` though).

**Review by drillprop — CHANGES_REQUESTED** [review]

File: src/components/CategoryVideos/CategoryVideos.tsx (hunk 1 above)
Comment (eldiegod) [reply src/components/CategoryVideos/CategoryVideos.tsx]: Not a real reason, it's just a visual component atm but functionality will be added later

File: src/hooks/useToggle.ts (hunk 6 above)
Comment (eldiegod) [reply src/hooks/useToggle.ts]: we do, for types

Comment (drillprop) [reply src/hooks/useToggle.ts]: Type inference works just fine here.  Are you sure? 
Without `as const`
image
With `as const`
image
Which is basically the same.

Anyway, why do we need this hook in the first place? It seems that it's not used at all.

Comment (eldiegod) [reply src/hooks/useToggle.ts]: not the same with `as const` first item of the array is a boolean and second item is a fn always; without `as const` type is an array of any size can have items booleans or fns.

But yes this might get removed or not depending on another PR that builds on this one

File: src/components/CategoryVideos/CategoryVideos.tsx (hunk 5 above)
Comment (eldiegod) [reply src/components/CategoryVideos/CategoryVideos.tsx]: I'll hold up on the searchbar changes as @kdembler mentioned it might get dropped

File: src/hooks/useToggle.ts (hunk 6 above)
Comment (drillprop) [reply src/hooks/useToggle.ts]: Just tested this. My mistake, you're right about `as const` 👍🏻

Comment (drillprop) [reply src/hooks/useToggle.ts]: basically without `as const` it's a regular array, and with `as const` it's a **tuple**(array with a fixed number of elements)

Comment (eldiegod) [reply src/hooks/useToggle.ts]: Yep, that the first element is type boolean and the second of type fn which is important.

Another way to do it is to define the return type of the fn but `as const` it's just convenient

**Comment by eldiegod** [comment]
Search bar was removed for now and will be added at some point in another PR

File: src/views/viewer/ChannelView/ChannelView.style.tsx [hunk 10]
```diff
@@ -4,12 +4,11 @@ import { fluidRange } from 'polished'
 import { ChannelLink } from '@/components/ChannelLink'
 import { TOP_NAVBAR_HEIGHT } from '@/components/TopbarBase'
 import { Button } from '@/shared/components/Button'
-import { IconButton } from '@/shared/components/IconButton'
+import { Search } from '@/shared/components/Search/Search'
```
Comment (kdembler) [inline src/views/viewer/ChannelView/ChannelView.style.tsx]: Should import from `@/shared/components/Search`. If `index.ts` doesn't exist there, let's add it

File: src/views/viewer/CategoryView/CategoryView.tsx [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import React from 'react'
+
+import { CategoryVideos } from '@/components/CategoryVideos/CategoryVideos'
```
Comment (kdembler) [inline src/views/viewer/CategoryView/CategoryView.tsx]: Should import from `index.ts`

File: src/views/viewer/index.ts [hunk 12]
```diff
@@ -5,3 +5,4 @@ export * from './NewView'
 export * from './PopularView'
 export * from './SearchOverlayView'
 export * from './VideoView'
+export * from './CategoryView/CategoryView'
```
Comment (kdembler) [inline src/views/viewer/index.ts]: Should re-export from `index.ts`

File: package.json [hunk 13]
```diff
@@ -51,6 +51,7 @@
     "@joystream/types": "0.16.1",
     "@loadable/component": "^5.15.0",
     "@polkadot/extension-dapp": "~0.37.3-17",
+    "@popperjs/core": "^2.10.1",
```
Comment (kdembler) [inline package.json]: I don't think this and `react-popper` are needed for this PR?

File: src/components/CategoryVideos/CategoryVideos.styles.tsx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+export const ControlsContainer = styled.div`
+  display: grid;
+  grid-template-columns: 1fr auto;
+  gap: ${sizes(4)};
+  align-items: center;
+  padding-bottom: ${sizes(4)};
+  border-bottom: 1px solid ${colors.gray[700]};
+
+  ${media.md} {
+    grid-template-columns: auto 168px 1fr 241px;
```
Comment (kdembler) [inline src/components/CategoryVideos/CategoryVideos.styles.tsx]: Where are those values coming from?

File: src/shared/components/Search/Search.tsx [hunk 15]
```diff
@@ -0,0 +1,171 @@
+import styled from '@emotion/styled'
```
Comment (kdembler) [inline src/shared/components/Search/Search.tsx]: I just rebased `discovery` with `dev`. Recently we've added some additional styles to the search component so this would need to be reflected here. Although I'm a bit unsure whether we should stick that into `Search` component. Design seems to define specific visuals just for the Channel search component. Maybe for now we should keep it in the channel view since we won't be using it here anyway?

**Review by kdembler — CHANGES_REQUESTED** [review]

File: package.json (hunk 13 above)
Comment (eldiegod) [reply package.json]: I removed it from popover-base will wait on it to get merged to rebase discovery here

File: src/components/CategoryVideos/CategoryVideos.styles.tsx (hunk 14 above)
Comment (eldiegod) [reply src/components/CategoryVideos/CategoryVideos.styles.tsx]: This is the size of the dropdown sections on MD breakpoint:
https://www.figma.com/file/UdBMxrLspjcf6IVYYJS8YP/[Page]-Discover-and-Video-category?node-id=252%3A93102

Comment (kdembler) [reply src/components/CategoryVideos/CategoryVideos.styles.tsx]: Could you help me understand what you mean by dropdown sections? Are you sure that Figma specifies this values as fixed as you currently do? I'm pretty sure the values depend on something else than a hardcoded value

Comment (eldiegod) [reply src/components/CategoryVideos/CategoryVideos.styles.tsx]: Maybe @toiletgranny might be able to clarify here, but the language dropdown and the sort by dropdown don't seem to be elements that scale or shrink in this MD breakpoint, in any case it was the searchbar the element that used to take free available space but is not there anymore

Comment (kdembler) [reply src/components/CategoryVideos/CategoryVideos.styles.tsx]: I still don't understand what these values are supposed to style. As far as I can tell, selects in the design just follow automatic width as all the other selects? I don't think they need to be explicitly sized like this?

File: src/views/viewer/ChannelView/ChannelView.style.tsx [hunk 16]
```diff
@@ -175,7 +175,7 @@ type SearchContainerProps = {
 export const SearchContainer = styled.div<SearchContainerProps>`
   display: flex;
   grid-area: search;
-  width: 100%;
+  width: 192px;
```
Comment (kdembler) [inline src/views/viewer/ChannelView/ChannelView.style.tsx]: Why are we making this change to channel view? Seems outside of scope of this PR

File: src/hooks/useToggle.ts [hunk 17]
```diff
@@ -0,0 +1,9 @@
+import { useCallback, useState } from 'react'
+
+export const useToggle = (initialState = false) => {
+  const [state, setState] = useState(initialState)
+
+  const toggle = useCallback(() => setState((state) => !state), [])
+
+  return [state, toggle] as const
+}
```
Comment (kdembler) [inline src/hooks/useToggle.ts]: Is this needed here? Let's remove it

File: package.json [hunk 18]
```diff
@@ -80,6 +81,7 @@
     "react-intersection-observer": "^8.31.0",
     "react-lottie-player": "^1.4.0",
     "react-number-format": "^4.4.4",
+    "react-popper": "^2.2.5",
```
Comment (kdembler) [inline package.json]: Let's remove those deps, they're not part of this PR, so they shouldn't be here.

**Review by kdembler — CHANGES_REQUESTED** [review]

File: src/views/viewer/ChannelView/ChannelView.style.tsx (hunk 16 above)
Comment (eldiegod) [reply src/views/viewer/ChannelView/ChannelView.style.tsx]: It's a fix I added in every PR after fixing the conflicts of the Search component, not ideal but it won't matter as it will be in discovery as well.

Comment (kdembler) [reply src/views/viewer/ChannelView/ChannelView.style.tsx]: I don't understand. Both this PR and discovery sprint don't make any changes to the `ChannelView` at all. What conflict do you mean it solves? There shouldn't be any conflicts

File: src/components/CategoryVideos/CategoryVideos.styles.tsx (hunk 14 above)
Comment (kdembler) [reply src/components/CategoryVideos/CategoryVideos.styles.tsx]: @eldiegod Still waiting on response for this

Comment (eldiegod) [reply src/components/CategoryVideos/CategoryVideos.styles.tsx]: They are the reserved space for the language dropdown and the sort by label + dropdown, I'm not sure what are you suggesting here instead? making them grow or shrink doesn't seem to fit design

File: src/views/viewer/ChannelView/ChannelView.style.tsx (hunk 16 above)
Comment (eldiegod) [reply src/views/viewer/ChannelView/ChannelView.style.tsx]: Yes, I think I was confused about this, removed the change

**Review by kdembler — APPROVED** [review]
